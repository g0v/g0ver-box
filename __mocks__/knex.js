import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
