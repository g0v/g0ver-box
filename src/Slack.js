import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
