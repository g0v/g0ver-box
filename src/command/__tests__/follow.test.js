import { client } from 'knex';
import Slack from '../../Slack';
import index from '../';

describe('follow command', () => {
  describe('follow <title>', () => {
    const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'follow g0ver box' };

    it('when project and event does not exist', async () => {
      client.mockReturnValueOnce([]);
      client.mockReturnValueOnce([]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Sorry! 找不到專案或活動',
      });

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });

    it('successfully follow when has event', async () => {
      client.mockReturnValueOnce([{ id: '5', title: 'g0ver box' }]);
      client.mockReturnValueOnce([]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Done, 追蹤 g0ver box 活動',
      });

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });

    it('successfully follow when has project', async () => {
      client.mockReturnValueOnce([]);
      client.mockReturnValueOnce([{ id: '5', title: 'g0ver box' }]);
      client.mockReturnValueOnce([]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Done, 追蹤 g0ver box 專案',
      });

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(3);
    });
  });

  describe('follow [title]', () => {
    const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'follow' };

    it('when project and event does not exist', async () => {
      client.mockReturnValueOnce([]);
      client.mockReturnValueOnce([]);

      await index(data);

      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Sorry! 找不到未追蹤的專案或活動',
      });

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });

    describe('when project or event is existed', () => {
      beforeEach(async () => {
        client.mockReturnValueOnce([{ id: '5', title: 'project title' }]);
        client.mockReturnValueOnce([{ id: '5', title: 'event title' }]);

        await index(data);
      });

      it('successfully follow project', async () => {
        client.mockReturnValueOnce([]);
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100',
          text: expect.stringContaining('請選擇要追蹤的專案或活動（輸入代碼）？\n`（輸入 exit 可離開）`'),
        });

        await index({ ...data, text: '1' });
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100',
          text: expect.stringContaining('Done, 追蹤 project title 活動'),
        });

        expect(client).toMatchSnapshot();
        expect(client).toHaveBeenCalledTimes(3);
      });

      it('successfully follow events', async () => {
        client.mockReturnValueOnce([]);
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100',
          text: expect.stringContaining('請選擇要追蹤的專案或活動（輸入代碼）？\n`（輸入 exit 可離開）`'),
        });

        await index({ ...data, text: '2' });
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100',
          text: expect.stringContaining('Done, 追蹤 event title 專案'),
        });

        expect(client).toMatchSnapshot();
        expect(client).toHaveBeenCalledTimes(3);
      });
    });
  });
});
