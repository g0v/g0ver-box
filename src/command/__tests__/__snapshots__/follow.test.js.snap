// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`follow command follow <title> successfully follow when has event 1`] = `
Array [
  "select * from event where datetime > NOW() and deleted_at is null and title = g0ver box order by datetime asc limit 1",
  "update event set follower_ids = array_append(array_remove(follower_ids, U03B2AB13), U03B2AB13) where deleted_at is null and id = 5",
]
`;

exports[`follow command follow <title> successfully follow when has project 1`] = `
Array [
  "select * from event where datetime > NOW() and deleted_at is null and title = g0ver box order by datetime asc limit 1",
  "select * from project where deleted_at is null and title = g0ver box limit 1",
  "update project set follower_ids = array_append(array_remove(follower_ids, U03B2AB13), U03B2AB13) where deleted_at is null and id = 5",
]
`;

exports[`follow command follow <title> when project and event does not exist 1`] = `
Array [
  "select * from event where datetime > NOW() and deleted_at is null and title = g0ver box order by datetime asc limit 1",
  "select * from project where deleted_at is null and title = g0ver box limit 1",
]
`;

exports[`follow command follow [title] when project and event does not exist 1`] = `
Array [
  "select * from event where follower_ids <> [U03B2AB13] and datetime > NOW() and deleted_at is null order by datetime asc limit 1000",
  "select * from project where follower_ids <> [U03B2AB13] and deleted_at is null limit 1000",
]
`;

exports[`follow command follow [title] when project or event is existed successfully follow events 1`] = `
Array [
  "select * from event where follower_ids <> [U03B2AB13] and datetime > NOW() and deleted_at is null order by datetime asc limit 1000",
  "select * from project where follower_ids <> [U03B2AB13] and deleted_at is null limit 1000",
  "update project set follower_ids = array_append(array_remove(follower_ids, U03B2AB13), U03B2AB13) where deleted_at is null and id = 5",
]
`;

exports[`follow command follow [title] when project or event is existed successfully follow project 1`] = `
Array [
  "select * from event where follower_ids <> [U03B2AB13] and datetime > NOW() and deleted_at is null order by datetime asc limit 1000",
  "select * from project where follower_ids <> [U03B2AB13] and deleted_at is null limit 1000",
  "update event set follower_ids = array_append(array_remove(follower_ids, U03B2AB13), U03B2AB13) where deleted_at is null and id = 5",
]
`;
